lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -529,19 +545,82 @@ getLegacyTransaction tid = do
     let inps = mempty
     pure $ Tx tid inps outs
 
+
+putWitness :: TxWitness -> Put
+putWitness (TxWitness bytes) = putByteString bytes
+
 putSignedTx :: [(TxIn, Coin)] -> [TxOut] -> [TxWitness] -> Put
 putSignedTx inputs outputs witnesses = do
-    putTx inputs outputs
+    putInputsOutputs inputs outputs
     unless (length inputs == length witnesses) $
         fail "number of witnesses must equal number of inputs"
     mapM_ putWitness witnesses
-  where
-    -- Assumes the `TxWitness` has been faithfully constructed
-    putWitness :: TxWitness -> Put
-    putWitness (TxWitness bytes) = putByteString bytes
 
-putTx :: [(TxIn, Coin)] -> [TxOut] -> Put
-putTx inputs outputs = do
+-- | Constructs a "normal" tx
+constructTransactionFragment
+    :: Monad m
+    => [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -> m (Hash "Tx", SealedTx)
+constructTransactionFragment ins outs mkWit = do
+    constructGenericSignedTx FragmentTransaction
+        mempty
+        ins outs mkWit
+        (const mempty)
+
+-- | Constructs a generic transaction of the form
+--
+-- 1. Payload
+-- 2. Inputs / outputs
+-- 3. Witnesses (including signed hash of 1 and 2)
+-- 4. Payload authentication (calculated using 1,2,3 as input)
+constructGenericSignedTx
+    :: Monad m
+    => FragmentSpec
+    -- ^ The kind of transaction
+    -> Put
+    -- ^ Payload (empty for normal utxo transactions)
+    -> [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -- ^ Function for producing witnesses for each address in the inputs.
+    -- E.g by looking up the private key and using it for signing with
+    -- @constructUtxoWitness@
+    -> (ByteString -> ByteString)
+    -- ^ Payload auth signing
+    -> m (Hash "Tx", SealedTx)
+constructGenericSignedTx spec putPayload ins outs lookupAddr auth = do
+    -- "Lookup the key" for each address, or fail.
+    mkWits <- mapM (\(_, TxOut addr _) -> lookupAddr addr) ins
+    return $ withHeader spec $
+        withAppendedSignature auth $ do
+            txH <- withHash $ do
+                putPayload
+                putInputsOutputs (map (second coin) ins) outs
+            forM_ mkWits $ \mkWit -> do
+                putWitness $ mkWit (Hash txH)
+  where
+    -- | Puts @x <> (sign x)@
+    withAppendedSignature
+        :: (ByteString -> ByteString)
+        -> PutM ()
+        -> PutM ()
+    withAppendedSignature sign x = do
+        sig <- sign <$> withBytes x
+        putByteString sig
+
+    withBytes :: PutM () -> PutM ByteString
+    withBytes putData = do
+        putData
+        let lookBehind = BL.toStrict $ runPut putData
+        return lookBehind
```

Comment:
My goal was to make `constructGenericSignedTx` as readable as possible.

With `withBytes`, (and related helpers) 

instead of:
```haskell
bs <- BL.toStrict $ runPut $ do
    putInputs ins
    putOutputs outs
putByteString bs
putSomething
let bsH = blake2b256 bs
-- do something with bsH
```

we can write:
```haskell
bsH <- blake2b256 <$> withBytes $ do
    putInputs ins
    putOutputs outs
putSomething
-- do something with bs
```

Thus avoid having to name & deal with the un-interesting intermediary values (here `bs`).

Note: not exact example form the code itself.

Reply:
I do find the first version much more readable than the second one :grimacing: 
It's using a vocabulary we already know and that is already quite clear in itself. Intermediate let bindings can help making things a bit clearer if it's really bothering you:

```hs 
    putByteString payload
    putSomething
  where
    payload = blake2b256 $ BL.toStrict $ runPut $ do
        putInputs ins
        putOutputs outs
```

But sticking with idioms and well-known library combinators make for more readable code. On a different aspect, this type of change should be done with consistency (i.e. across a whole module or package). It really hurts maintainability to run into functions that are written using different style and approaches, especially within a same module. 

Whenever you modify or extend an existing module, stick as much as possible with the existing style even if you disagree with it. If really something needs to change, then it should be done consistently across the whole module. Yet, that's not something we typically do in sprints.

Reply:
Some options:
```haskell
        withAppendedSignature auth $ do
            txH <- withHash $ do
                putPayload
                putInputsOutputs (map (second coin) ins) outs
            forM_ mkWits $ \mkWit -> do
                putWitness $ mkWit (Hash txH)
```
```haskell
        auth <- mkAuth <$> withBytes $ do
            txH <- blake2b256 <$> withBytes $ do
                putPayload
                putInputsOutputs (map (second coin) ins) outs
            forM_ mkWits $ \mkWit -> do
                putWitness $ mkWit (Hash txH)
        putByteString auth
```

```haskell
        let payIOs = runPut' $ do
                putPayload
                putInputsOutputs (map (second coin) ins) outs
        let wits = runPut' $ do
                forM_ mkWits $ \mkWit -> do
                    putWitness $ mkWit (Hash $ blake2b256 payIOs)
        let payIOsAndWits = payIOs <> wits
        putByteString payIOsAndWits
        putByteString (mkAuth payIOsAndWits)
```